### PR TITLE
HSD8-445 Added alter to conslidated date field sorting

### PR DESCRIPTION
**Pull request by pookmish**
# READY FOR REVIEW

# Summary
- Allow for multiple date fields to be used as a single sort.

# Needed By (Date)
- Soon

# Urgency
- low

# Steps to Test

1. checkout this branch
1. clear cache
1. create some events and news items (keep track of what dates are on them) alternating dates between the two content types.
1. create a new view
1. add some fields to the view and add the two date fields as sorting
1. validate the view sorts the items togeather.
1. optionally add another sort to the bottom or the top of the date fields, validate it sorts correctly still.

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

File: docroot/modules/humsci/hs_field_helpers/src/Plugin/views/query/Sql.php [hunk 1]
```diff
@@ -0,0 +1,122 @@
+<?php
+
+namespace Drupal\hs_field_helpers\Plugin\views\query;
+
+use Drupal\field\Entity\FieldStorageConfig;
+use Drupal\views\Plugin\views\query\Sql as OriginalSql;
+use Drupal\views\ViewExecutable;
+
+/**
+ * Class Sql that adds extra functionality to views Sql.
```
Comment (sherakama) [inline docroot/modules/humsci/hs_field_helpers/src/Plugin/views/query/Sql.php]: What kind of extra functionality?

Comment (pookmish) [reply docroot/modules/humsci/hs_field_helpers/src/Plugin/views/query/Sql.php]: the secretive kind.

File: docroot/modules/humsci/hs_field_helpers/src/Plugin/views/query/Sql.php [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+    $expression = [];
+    $orderby_delta = -1;
+
+    // Build the expression.
+    foreach ($date_fields as $field_alias => $direction) {
+      $field = $this->fields[$field_alias];
+
+      // Set null values to 0.
+      $expression[] = 'COALESCE(' . $field['table'] . '.' . $field['field'] . ', 0)';
```
Comment (sherakama) [inline docroot/modules/humsci/hs_field_helpers/src/Plugin/views/query/Sql.php:66]: The `COALESCE` function seems to be valid for PGSQL, MySQL, and MSSQL. Not sure if there are slight behavior differences.

File: docroot/modules/humsci/hs_field_helpers/src/Plugin/views/query/Sql.php [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+   */
+  protected function getDateFields() {
+    $field_types = ['datetime', 'daterange'];
+    $date_fields = [];
+
+    foreach ($this->orderby as $order_item) {
+      $field_alias = $order_item['field'];
+      $field_data = $this->fields[$field_alias];
+      $table = $field_data['table'];
+      if (strpos($table, '__') === FALSE) {
```
Comment (sherakama) [inline docroot/modules/humsci/hs_field_helpers/src/Plugin/views/query/Sql.php:112]: Why this check?

File: docroot/modules/humsci/hs_field_helpers/src/Plugin/views/query/Sql.php [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        continue;
+      }
+      list($entity_type, $field_name) = explode('__', $table);
```
Comment (sherakama) [inline docroot/modules/humsci/hs_field_helpers/src/Plugin/views/query/Sql.php:116]: http://php.net/manual/en/function.list.php

The list function's behavior is different in PHP5 vs PHP7. Is there a way you can ensure that this module is dependant on PHP7 or protect against running backward on PHP5.6

Comment (pookmish) [reply docroot/modules/humsci/hs_field_helpers/src/Plugin/views/query/Sql.php:116]: I read that warning and i dont believe it applies to this code. I tried out different versions using http://sandbox.onlinephpfunctions.com/
php 5 and php 7 both produce the same result when exploding a string.

File: docroot/modules/humsci/hs_field_helpers/src/Plugin/views/query/Sql.php (hunk 3 above)
Comment (pookmish) [reply docroot/modules/humsci/hs_field_helpers/src/Plugin/views/query/Sql.php:112]: comment added. it basically forces us to only operate on CCK fields. Base fields like node title and items arent applicable.

File: docroot/modules/humsci/hs_field_helpers/src/Plugin/views/query/Sql.php (hunk 2 above)
Comment (pookmish) [reply docroot/modules/humsci/hs_field_helpers/src/Plugin/views/query/Sql.php:66]: just noting?

File: docroot/modules/humsci/hs_field_helpers/src/Plugin/views/query/Sql.php (hunk 1 above)
Comment (pookmish) [reply docroot/modules/humsci/hs_field_helpers/src/Plugin/views/query/Sql.php]: Greatest and least expressions

**Review by boznik — APPROVED** [review]
Working as expected

**Pull request merged**